ext/soap/php_wss.c

```diff
@@ -0,0 +1,279 @@
+						#endif
+						result = zend_string_init(buffer_content, strlen(buffer_content), 0);
+					}
+				}
+			}
+
+			xmlXPathFreeObject(xpath_obj);
+			xmlXPathFreeContext(xpath_ctx);
+		}
+		xmlOutputBufferClose(buffer);
+	}
+
+	return result;
+}
+/* }}} */
+
+static int add_reference_to_signed_info(zval *this_ptr, xmlNodePtr signed_info, xmlNodePtr source, zend_string *source_uri, xmlNsPtr ns_ds) /* {{{ */
+{
+	xmlNodePtr reference = xmlNewChild(signed_info, ns_ds, BAD_CAST("Reference"), NULL);
+	xmlSetProp(reference, BAD_CAST("URI"), BAD_CAST(ZSTR_VAL(source_uri)));
+	xmlNodePtr reference_trsfrms = xmlNewChild(reference, ns_ds, BAD_CAST("Transforms"), NULL);
+	xmlNodePtr reference_trsfrm = xmlNewChild(reference_trsfrms, ns_ds, BAD_CAST("Transform"), NULL);
+	xmlSetProp(reference_trsfrm, BAD_CAST("Algorithm"), BAD_CAST(SOAP_WSS_EXC_C14N));
+	xmlNodePtr reference_dm = xmlNewChild(reference, ns_ds, BAD_CAST("DigestMethod"), NULL);
+	
+	const php_hash_ops *ops = NULL;
+	zend_string *algo = NULL;
+	zend_long digest_method = Z_LVAL_P(Z_CLIENT_WSS_DIGEST_METHOD_P(this_ptr));
+	switch (digest_method) {
+		case SOAP_WSS_DIGEST_METHOD_SHA1:			
+			algo = zend_string_init("sha1", strlen("sha1"), 0);
+			ops = php_hash_fetch_ops(algo);
+			xmlSetProp(reference_dm, BAD_CAST("Algorithm"), BAD_CAST(SOAP_WSS_SHA1));
+			break;
+		case SOAP_WSS_DIGEST_METHOD_SHA256:
+			algo = zend_string_init("sha256", strlen("sha256"), 0);
+			ops = php_hash_fetch_ops(algo);
+			xmlSetProp(reference_dm, BAD_CAST("Algorithm"), BAD_CAST(SOAP_WSS_SHA256));
+			break;
+		case SOAP_WSS_DIGEST_METHOD_SHA512:
+			algo = zend_string_init("sha512", strlen("sha512"), 0);
+			ops = php_hash_fetch_ops(algo);
+			xmlSetProp(reference_dm, BAD_CAST("Algorithm"), BAD_CAST(SOAP_WSS_SHA512));			
+			break;
+	}
+
+	if (ops == NULL) {
+		if (algo != NULL) {
+			php_error_docref(NULL, E_WARNING, "invalid hashing algorithm, %s", ZSTR_VAL(algo));
+			zend_string_release_ex(algo, 0);
+		} else {
+			php_error_docref(NULL, E_WARNING, "unknown hashing algorithm");
+		}
+		return 0;
+	}
+
+	zend_string_release_ex(algo, 0);
+
+	zend_string *exc_c14n = get_exc_c14n_from_node_ptr(source);
+	if (exc_c14n != NULL) {
+		void *context = php_hash_alloc_context(ops);
+		ops->hash_init(context, NULL);
+		ops->hash_update(context, (unsigned char *)ZSTR_VAL(exc_c14n), ZSTR_LEN(exc_c14n));
+		
+		zend_string *digest = zend_string_alloc(ops->digest_size, 0);
+		ops->hash_final((unsigned char *)ZSTR_VAL(digest), context);
+		efree(context);
+
+		ZSTR_VAL(digest)[ops->digest_size] = 0;
+
+		zend_string *digest_base64 = php_base64_encode((unsigned char *)ZSTR_VAL(digest), ZSTR_LEN(digest));
+		xmlNewChild(reference, ns_ds, BAD_CAST("DigestValue"), BAD_CAST(ZSTR_VAL(digest_base64)));
+
+		zend_string_release(digest_base64);
+		zend_string_release(digest);
+		zend_string_release_ex(exc_c14n, 0);
+
+		return 1;
+	} else {
+		php_error_docref(NULL, E_WARNING, "exclusive C14N failed, URI \"%s\"", ZSTR_VAL(source_uri));
```

Comment:
This should throw.

Reply:
I think we want to keep it consistent and the extension mainly uses warnings from the quick look so I would keep ti that way. The algorithm above could be exception but this is failure so it might be better to be a warning for consistency.

Reply:
The problem is that unless one sets an error handler, there is no way to know what happened.

Reply:
SOAP extension seem to use E_ERROR's, E_WARNING's & SoapFault object to signal error events. So maybe some of these new warnings (**"invalid hashing algorithm"** & **"unknown hashing algorithm"**) could be E_ERROR's instead and this one maybe a SoapFault?

Reply:
E_ERROR and E_WARNING are bad.  No idea about SoapFault.